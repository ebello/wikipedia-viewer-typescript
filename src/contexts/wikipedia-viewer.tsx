import React, { useState, useEffect, ReactNode } from 'react';
import PropTypes from 'prop-types';

interface ViewingHistoryItem {
  timestamp: number,
  title: string,
}

interface WikipediaContextValue {
  readonly viewingHistory: ViewingHistoryItem[],
  addToViewingHistory: (title: string) => void,
  clearViewingHistory: () => void,
}

type ContextProps = {
  children: ReactNode,
};

const WikipediaViewerContext = React.createContext<WikipediaContextValue>({
  viewingHistory: [],
  addToViewingHistory: () => null,
  clearViewingHistory: () => null,
});

const WikipediaViewer = ({ children }: ContextProps): JSX.Element => {
  const viewingHistoryStr = window?.localStorage.getItem('viewingHistory');
  const defaultHistory = JSON.parse(viewingHistoryStr || '') as ViewingHistoryItem[] || [];
  const [viewingHistory, setViewingHistory] = useState(defaultHistory);

  const addToViewingHistory = (title: string) => {
    setViewingHistory([...viewingHistory, {
      timestamp: Date.now(),
      title,
    }]);
  };

  const clearViewingHistory = () => setViewingHistory([]);

  useEffect(() => {
    window.localStorage.setItem('viewingHistory', JSON.stringify(viewingHistory));
  }, [viewingHistory]);

  return (
    <WikipediaViewerContext.Provider
      value={{
        viewingHistory,
        addToViewingHistory,
        clearViewingHistory,
      }}
    >
      {children}
    </WikipediaViewerContext.Provider>
  );
};

WikipediaViewer.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default WikipediaViewer;

export { WikipediaViewerContext };
