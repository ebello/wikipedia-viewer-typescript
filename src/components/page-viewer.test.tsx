import React from 'react';
import {
  render, screen, waitFor, sleep,
} from '../test-utils';
import { WikipediaViewerContext } from '../contexts/wikipedia-viewer';
import PageViewer from './page-viewer';

describe('Page Viewer', () => {
  it('renders', async () => {
    const { asFragment } = render(<PageViewer title="testing" />);
    expect(await screen.findByRole('heading')).toHaveTextContent('testing');
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls addToViewingHistory', async () => {
    const addToViewingHistory = jest.fn();
    render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [],
          addToViewingHistory,
          clearViewingHistory: () => {},
        }}
      >
        <PageViewer title="testing" />
      </WikipediaViewerContext.Provider>,
    );
    await waitFor(() => expect(addToViewingHistory).toHaveBeenCalled());
  });

  it('does not call addToViewingHistory if unmounted', async () => {
    const addToViewingHistory = jest.fn();
    const { unmount } = render(
      <WikipediaViewerContext.Provider
        value={{
          viewingHistory: [],
          addToViewingHistory,
          clearViewingHistory: () => {},
        }}
      >
        <PageViewer title="testing" />
      </WikipediaViewerContext.Provider>,
    );
    unmount();
    await sleep(1000);
    expect(addToViewingHistory).not.toHaveBeenCalled();
  });
});
