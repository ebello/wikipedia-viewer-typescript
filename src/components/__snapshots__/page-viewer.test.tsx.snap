// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Viewer renders 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 1.3;
  margin-top: 0;
  margin-bottom: 16px;
  font-size: 24px;
}

@media screen and (min-width:46.25em) {
  .emotion-0 {
    font-size: 32px;
  }
}

<h1
    class="emotion-0"
    role="heading"
  >
    testing
  </h1>
  <div>
    <p>
      body text
    </p>
  </div>
</DocumentFragment>
`;
